### ISSUE TITLE
Test suite crashes depending on sequence of tests

### ISSUE THREAD

**Issue by scotws** [scotws/TaliForth2#76]
So, now it is reproducible. Depending on the sequence of tests, the suite crashes at the same place:
```
: ifsym  compiled
   [ -3 2 / -1 = invert ] literal if postpone \ then ;  ok
  ok
\ the system might do either floored or symmetric division.  ok
\ since we have already tested m*, fm/mod, and sm/rem we can use them in test.  ok
iffloored : t/mod  >r s>d r> fm/mod ; 
        PC  AC XR YR SP NV-BDIZC
65C02: 0007 00 78 fb f8 00110110
```
This only happens when the `tali` group of tests follows `double` and `string`. Specifically:
- double string tali -> **CRASH**
- tali -> *fine*
- string tali -> *fine*
- double tali -> *fine*
- string -> *fine*
- double -> *fine*
- double string -> *fine*
- double tali string -> *fine*
- string double tali -> **CRASH**
- tali string double -> *fine*

Note calling the test suite with no explicit sequence leaves it up to the Python script, and that can be different each time. 

For the moment, am running tests with the last sequence until this bug is taken care of.

**Comment by scotws** [comment]
Note that the "tali" set of words will be renamed "core".

**Comment by SamCoVT** [comment]
That's very interesting.  I'll see if I can reproduce that tonight.

**Comment by SamCoVT** [comment]
- talitest.py -t double string core -> crashes on:
`iffloored : t/mod  >r s>d r> fm/mod ;` where iffloored is:
```
: iffloored
   [ -3 2 / -2 = invert ] literal if postpone \ then ;
```
It looks like the iffloored is checking (immediately) to see if the result of divison is floored or symmetric around zero, and then commenting out that line if its the wrong type of rounding/truncating.  I totally didn't know you could postpone \ for later commenting things out.  These forth folks are crazy.

The good news is that I can exactly reproduce your crash.  The interesting part here is the $0007 in PC at the moment of the crash, as well as the fact the the order of the tests matters.

Also good news is the fact that when fed the tester.fs, double.fs, string.fs and core.fs in that order, my hardware runs the above test perfectly, but it did reset all by itself right at:
```
: gp5  compiled
   base @ <true>  compiled
   max-base 1+ 2 do   \ for each possible base  compiled
      i base !    \ tbd: assumes base works  compiled
      i 0 <# #s #> s" 10" s= and  compiled
    swap base ! ;  ok
{ gp5 -> <true> } 
Tali Forth 2 for the 65c02
Version ALPHA 17. June 2018
```
I'll keep digging to see what I can find.

**Comment by SamCoVT** [comment]
I was able to get it to run most of the tests without stopping on my hardware.

I noticed, while it was running, that the further along it got, the more likely it was to drop characters, especially near the beginning of the line.  I wonder if it's taking longer and longer to look up the words. Here's an example where it's especially bad, but it seems to recover after this:
```
 : gp1  <# 41 hold 42 hold 0 0 #> s" BA" s= ;  ok
{-> <true> } >{->< Undefined word
  ok
: gp2  <# -1 sign 0 sign -1 sign 0 0 #> s" --" s= ;  ok
{rue> } >{rue>< Undefined word
  ok
: gp3  <# 1 0 # # #> s" 01" s= ;  ok
{p3 -> <true> } >{p3< Undefined word
  ok
: gp4  <# 1 0 #s #> s" 1" s= ;  ok
{ gp4 -> <true> }  ok
  ok
```

With the dictionary grown by the other tests, it could take longer to find built-in words, which is mostly what core.fs is testing.  On my hardware, while it's trying to run the previous line, the ACIA gets (and holds) the first character of the next line, but it only has a 1 byte buffer, so the next chars are dropped until Tali gets around to ACCEPT again.

Adding intercharacter delays made the crash go away and most of the tests ran with just a few tests mangled as shown above.  It gacked on the CASE tests, which makes sense - I haven't reassembled my hardware branch since then.  I'll rebase to get those changes and try it again on the hardware with even longer intercharacter and interline delays (set in minicom, and I just paste the tests over to it.) and see how that goes.

**Comment by SamCoVT** [comment]
I ran the tests in the emulator with a 20ms delay between chars, and it still crashed in the exact same way.

I ran the tests on the hardare with a 10ms delay between chars and a 500ms delay between lines.  It didn't drop any tests this time, but it *DID* crash in the exact same spot as the emulator.  Yay! and :( at the same time.

I added a `see iffloored` just before the crash, and also changed the crashing line to `iffloored : 0 ;` to make sure it's the iffloored.  I got:
```
see iffloored 
 nt: 24E6  xt: 24F7  NN 
 size (decimal): 18 
 
24F7  20 D5 8C FF FF 20 1A 9A  09 25 A5 0A 85 0C A5 0B 
2507  85 0D 

24F7  jsr 8CD5
24FA  bbs7 20FF
24FD  inc.a
24FE  txs
24FF  ora.# 25
2501  lda.z 0A
2503  sta.z 0C
2505  lda.z 0B
2507  sta.z 0D
 ok
  ok
iffloored : 0 ; 
        PC  AC XR YR SP NV-BDIZC
65C02: 0007 00 78 fb f8 00110110

Py65 Monitor
```
The disassembly from see isn't useful past the jsr for a constant because the constant ($ffff)  We'll just take it right from the hex:

`20 D5 8C` is a call to literal_runtime and the `FF FF` that comes next is what to put on the stack.
`20 1A 9A` is a call to zero_branch_runtime and `09 25` is the location to jump to, which is the rts (not shown in see) of this word.  We won't take that jump, however, because of the $ffff on the stack.
`A5 0A` is LDA ciblen
`85 0C` is STA toin
`A5 0B` is LDA ciblen+1
`85 0D` is STA toin+1
That should "use up" the rest of the input by moving toin to the end of the buffer.
That looks reasonable.  

I've tracked it all the way down to PARSE-NAME and it looks like PARSE-NAME subtracts toin from ciblen to see how many chars are left, but then it subtracts 1 because one is a length and the other is a 0-based offset.  The issue, at this point, is this will calculate that there are $ffff characters left to scan.  I'll try adding a check for toin = ciblen (really for the subtraction result to be zero) and branch down to _empty_line if that's the case.  That seems to work.  Pull request #79 should fix it.
